internal/buildsign/pod/manager.go

```diff
@@ -0,0 +1,133 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/api"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/buildsign"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/kernel"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+)
+
+type podManager struct {
+	client              client.Client
+	maker               Maker
+	signer              Signer
+	buildSignPodManager BuildSignPodManager
+}
+
+func NewManager(client client.Client, maker Maker, signer Signer, buildSignPodManager BuildSignPodManager) buildsign.Manager {
+	return &podManager{
+		client:              client,
+		maker:               maker,
+		signer:              signer,
+		buildSignPodManager: buildSignPodManager,
+	}
+}
+
+func (pm *podManager) GetStatus(ctx context.Context, name, namespace, kernelVersion string,
+	action kmmv1beta1.BuildOrSignAction, owner metav1.Object) (kmmv1beta1.BuildOrSignStatus, error) {
+	podType := PodTypeBuild
+	if action == kmmv1beta1.SignImage {
+		podType = PodTypeSign
+	}
+	normalizedKernel := kernel.NormalizeVersion(kernelVersion)
+	foundPod, err := pm.buildSignPodManager.GetModulePodByKernel(ctx, name, namespace, normalizedKernel, podType, owner)
+	if err != nil {
+		if !errors.Is(err, ErrNoMatchingPod) {
+			return kmmv1beta1.BuildOrSignStatus(""), fmt.Errorf("failed to get pod %s/%s, action %s: %v", namespace, name, action, err)
+		}
+		return kmmv1beta1.BuildOrSignStatus(""), nil
+	}
+	status, err := pm.buildSignPodManager.GetPodStatus(foundPod)
+	if err != nil {
+		return kmmv1beta1.BuildOrSignStatus(""), fmt.Errorf("failed to get status from the pod %s/%s, action %s: %v",
+			foundPod.Namespace, foundPod.Name, action, err)
+	}
+	switch status {
+	case StatusCompleted:
+		return kmmv1beta1.ActionSuccess, nil
+	case StatusFailed:
+		return kmmv1beta1.ActionFailure, nil
+	}
+	return kmmv1beta1.BuildOrSignStatus(""), nil
+}
+
+func (pm *podManager) Sync(ctx context.Context, mld *api.ModuleLoaderData, pushImage bool, owner metav1.Object, action kmmv1beta1.BuildOrSignAction) error {
+	logger := log.FromContext(ctx)
+
+	var podType string
+	var podTemplate *v1.Pod
+	var err error
+	if action == kmmv1beta1.BuildImage {
+		logger.Info("Building in-cluster")
+		podType = PodTypeBuild
+		podTemplate, err = pm.maker.MakePodTemplate(ctx, mld, owner, pushImage)
+	} else {
+		logger.Info("Signing in-cluster")
+		podType = PodTypeSign
+		podTemplate, err = pm.signer.MakePodTemplate(ctx, mld, owner, pushImage)
+	}
+	if err != nil {
+		return fmt.Errorf("could not make Pod template: %v", err)
+	}
+
+	p, err := pm.buildSignPodManager.GetModulePodByKernel(ctx, mld.Name, mld.Namespace,
+		mld.KernelNormalizedVersion, podType, owner)
+
+	if err != nil {
+		if !errors.Is(err, ErrNoMatchingPod) {
+			return fmt.Errorf("error getting the %s pod: %v", podType, err)
+		}
```

Comment:
This is the second time we are running this code (first one was in the previous method). How about we move the check to see if the error is `ErrNoMatchingPod` inside the `GetModulePodByKernel`)?

Reply:
ErrNoMatchingPod is actually returned by the GetModulePodByKernel on purpose, so that the callee code will know what to do.